src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php

```diff
@@ -60,86 +110,97 @@ public function buildForm(FormBuilderInterface $builder, array $options)
 
     public function configureOptions(OptionsResolver $resolver)
     {
-        $choiceListCache = &$this->choiceListCache;
         $registry = $this->registry;
-        $propertyAccessor = $this->propertyAccessor;
+        $choiceListFactory = $this->choiceListFactory;
+        $idReaders = &$this->idReaders;
+        $choiceLoaders = &$this->choiceLoaders;
         $type = $this;
 
-        $loader = function (Options $options) use ($type) {
-            $queryBuilder = (null !== $options['query_builder'])
-                ? $options['query_builder']
-                : $options['em']->getRepository($options['class'])->createQueryBuilder('e');
+        $choiceLoader = function (Options $options) use ($choiceListFactory, &$choiceLoaders, $type) {
+            // This closure and the "query_builder" options should be pushed to
+            // EntityType in Symfony 3.0 as they are specific to the ORM
+
+            // Unless the choices are given explicitly, load them on demand
+            if (null === $options['choices']) {
+                // We consider two query builders with an equal SQL string and
+                // equal parameters to be equal
+                $qbParts = $options['query_builder']
+                    ? array(
+                        $options['query_builder']->getQuery()->getSQL(),
+                        $options['query_builder']->getParameters()->toArray(),
```

Comment:
Yes, the hash generation function traverses the array and replaces objects by their hashes.